shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/ManagerSyncServiceImpl.java

```diff
@@ -176,24 +181,34 @@ private void processingComplete(Class<?> entity) {
      * @throws IllegalStateException in case of an invalid HTTP response
      */
     private void validateHttpResponseBeforeParsing(HttpResponse<Buffer> httpResponse) {
+        if (httpResponse.statusCode() == 401) {
+            LOGGER.debug("Request to the manager was unauthorized.");
+            threadContext.contextualRunnable(this::refreshTokens);
+        }
         if (httpResponse.statusCode() < 200 || httpResponse.statusCode() >= 400) {
             throw new DeserializationException(String.format("Got %d HTTP status code response, skipping deserialization process", httpResponse.statusCode()));
         }
     }
 
     private Uni<HttpResponse<Buffer>> getAuthenticatedRequest(HttpRequest<Buffer> request, Function<HttpRequest<Buffer>, Uni<HttpResponse<Buffer>>> executor) {
-        Tokens tokens = currentTokens;
-        if (tokens.isAccessTokenExpired()) {
+        if (currentTokens.isAccessTokenExpired() || currentTokens.isAccessTokenWithinRefreshInterval()) {
             LOGGER.debug("Shard authentication token has expired");
-            try {
-                tokens = client.refreshTokens(tokens.getRefreshToken()).await().atMost(SSO_CONNECTION_TIMEOUT);
-            } catch (RuntimeException e) {
-                LOGGER.warn("Shard could not fetch a new authentication token from sso server.");
-                throw e;
-            }
-            currentTokens = tokens;
+            refreshTokens();
         }
         request.bearerTokenAuthentication(currentTokens.getAccessToken());
         return executor.apply(request);
     }
+
+    private void refreshTokens() {
+        LOGGER.info("Refreshing tokens..");
+        Tokens tokens = currentTokens;
+        try {
+            tokens = client.refreshTokens(tokens.getRefreshToken()).await().atMost(SSO_CONNECTION_TIMEOUT);
+        } catch (OidcClientException e) {
+            LOGGER.warn("Shard could not use refresh token. Trying to get a new fresh token.", e);
```

Comment:
Do we know when this can happen? Are there expected scenarios where this is supposed to happen?

Reply:
The refresh token has a lifespan: when it is not valid anymore we have to retrieve a brand new token with its refresh token